### PR TITLE
Adds custom-http-errors per-ingress

### PR DESCRIPTION
**Pull request by jasongwartz**

### CODE REVIEW

File: rootfs/etc/nginx/template/nginx.tmpl [hunk 1]
```diff
@@ -794,6 +794,33 @@ stream {
         {{ $server.ServerSnippet }}
         {{ end }}
 
+        {{ range $errCode := $server.AllCustomHTTPErrors }}
```
Comment (jasongwartz) [inline rootfs/etc/nginx/template/nginx.tmpl]: This still needs to be deduplicated somehow with the template `CUSTOM_ERRORS` above in this file.

**Comment by jasongwartz** [comment]
@Globegitter if you want to try spinning this up locally, you can follow the directions here: https://github.com/kubernetes/ingress-nginx/blob/29c5d770688b04d0a8beedf70aebd76990332d56/docs/development.md - ie `make dev-env`

I also had to manually mount the source directory into the minikube vm like so: `minikube mount $GOPATH/src/k8s.io/ingress-nginx:/home/$USER/go/src/k8s.io/ingress-nginx/`

File: docs/user-guide/nginx-configuration/annotations.md [hunk 2]
```diff
@@ -213,6 +214,15 @@ The ingress controller requires a [default backend](../default-backend.md).
 This service handles the response when the service in the Ingress rule does not have endpoints.
 This is a global configuration for the ingress controller. In some cases could be required to return a custom content or format. In this scenario we can use the annotation `nginx.ingress.kubernetes.io/default-backend: <svc name>` to specify a custom default backend.
 
+### Custom HTTP Errors
+
+Like the ConfigMap `custom-http-errors`, this annotation will set NGINX `proxy-intercept-errors`, but only for the hostname associated with this ingress. Different ingresses can specify different sets of error codes. If multiple ingress objects share the same hostname, this annotation can be used to intercept different error codes for each hostname. If `custom-http-errors` is also specified globally, the error values specified in this annotation will override the global value for this ingress' hostname and path.
```
Comment (Globegitter) [inline docs/user-guide/nginx-configuration/annotations.md]: Not sure I follow this sentence: `If multiple ingress objects share the same hostname, this annotation can be used to intercept different error codes for each hostname. `. Maybe you can explain in different words?

Comment (jasongwartz) [reply docs/user-guide/nginx-configuration/annotations.md]: ah typo - should say `can be used to intercept different error codes for each path`.

Effectively, if there are two different ingresses for the same host but different paths, each ingress can specify its own set of `custom-http-errors` (which then applies for the corresponding path of that ingress).

eg:
An ingress object for api.company.com/v1 can override "500,502,503", and a different ingress object for api.company.com/v2 can override "401,503,504" etc, and that will be respected.

Comment (jasongwartz) [reply docs/user-guide/nginx-configuration/annotations.md]: I took another crack at it, let me know if it's better or worse.

File: rootfs/etc/nginx/template/nginx.tmpl [hunk 3]
```diff
@@ -1149,6 +1151,15 @@ stream {
...
             {{ end }}
 
+            {{/* if a location-specific error override is set, add the proxy_intercept here */}}
+            {{ if $location.CustomHTTPErrors }}
+            # Custom error pages per ingress
+            proxy_intercept_errors on;
+            {{ end }}
+
+            {{ range $errCode := $location.CustomHTTPErrors }}
+            error_page {{ $errCode }} = @custom_{{ $errCode }};{{ end }}
```
Comment (Globegitter) [inline rootfs/etc/nginx/template/nginx.tmpl]: Can you move the `{{ end }}` to the next line.

**Comment by Globegitter** [comment]
Maybe you can make that PR against another branch - I think it is nice if we can keep this master in sync with the upstream master. Merging this PR into master makes it more difficult.

File: rootfs/etc/nginx/template/nginx.tmpl (hunk 3 above)
Comment (jasongwartz) [reply rootfs/etc/nginx/template/nginx.tmpl]: As mentioned in person, this is consistent with other {{ range }} blocks in this template file.

**Comment by jasongwartz** [comment]
@Globegitter absolutely, I just opened this for you to look at, but intend to close this and open a fresh PR upstream.

**Pull request closed without merging**